### PR TITLE
8309171: Test vmTestbase/nsk/jvmti/scenarios/jni_interception/JI05/ji05t001/TestDescription.java fails after JDK-8308341

### PR DESCRIPTION
**Pull request by dholmes-ora**
The fix for [JDK-8308341](https://bugs.openjdk.org/browse/JDK-8308341) overlooked the fact that during VM initialization agents can run and create native threads that will attach to (and potentially detach from) the VM. The check that VM init was complete before allowing those operations was too strong and has to be reverted. All the guards, except for that of GetCreatedJavaVMs itself, is reverted to the way it was before JDK-8308341.

Testing: tiers 1-4 (in progress)

Thanks.

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8309171](https://bugs.openjdk.org/browse/JDK-8309171): Test vmTestbase/nsk/jvmti/scenarios/jni_interception/JI05/ji05t001/TestDescription.java fails after JDK-8308341

### Reviewers
 * [Daniel D. Daugherty](https://openjdk.org/census#dcubed) (@dcubed-ojdk - **Reviewer**) ⚠️ Review applies to [3d8378bb](https://git.openjdk.org/jdk/pull/14239/files/3d8378bbe96b7438caa4d1e0b1ccad40ae74dad6)
 * [Gerard Ziemski](https://openjdk.org/census#gziemski) (@gerard-ziemski - Committer) ⚠️ Review applies to [3d8378bb](https://git.openjdk.org/jdk/pull/14239/files/3d8378bbe96b7438caa4d1e0b1ccad40ae74dad6)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/14239/head:pull/14239` \
`$ git checkout pull/14239`

Update a local copy of the PR: \
`$ git checkout pull/14239` \
`$ git pull https://git.openjdk.org/jdk.git pull/14239/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 14239`

View PR using the GUI difftool: \
`$ git pr show -t 14239`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/14239.diff">https://git.openjdk.org/jdk/pull/14239.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/14239#issuecomment-1569590007)

### CODE REVIEW

**Review by dcubed-ojdk — APPROVED** [review]
Thumbs up. The fix appears to do exactly what you described in the bug report.
Hopefully your in process testing goes well.

**Comment by azvegint** [comment]
Please note that the associated test is now in the problem list, see #14250

**Comment by dcubed-ojdk** [comment]
vmTestbase/nsk/jvmti/scenarios/jni_interception/JI05/ji05t001/TestDescription.java
has been ProblemListed. Please update this PR and restore the previous entry for
vmTestbase/nsk/jvmti/scenarios/jni_interception/JI05/ji05t001/TestDescription.java
before integrating this fix.

The previous entry looked like this:
```
vmTestbase/nsk/jvmti/scenarios/jni_interception/JI05/ji05t001/TestDescription.java 8219652 aix-ppc64
```

**Comment by AlanBateman** [comment]
I skimmed through the JVMTI spec and I don't see anything to set expectations that a thread can potentially JNI attach before the "live phase".  However, once the VM is in the live phase and the VMInit callback is invoked then it's game on. This makes me wonder if VM_Creation_State needs an additional state between IN_PROGRESS and CREATED to indicate that the VM is initialized so  IN_PROGRESS -> "new-state"  in Threads::create_vm and before JvmtiExport::post_vm_initialized, and "nw-state" -> CREATED when Threads::create_vm completes. That would prevent JNI attach before entering the live phase.

**Review by gerard-ziemski — APPROVED** [review]
Looks to me like the places you touched here are equivalent to the way it used to be originally.

**Comment by gerard-ziemski** [comment]
If we could lock the VM init section, then any thread trying to do anything that depended on VM to be initialized, would just sit waiting till the VM is ready...

File: src/hotspot/share/prims/jni.cpp [hunk 1]
```diff
@@ -3991,7 +3989,7 @@ jint JNICALL jni_GetEnv(JavaVM *vm, void **penv, jint version) {
 
 jint JNICALL jni_AttachCurrentThreadAsDaemon(JavaVM *vm, void **penv, void *_args) {
   HOTSPOT_JNI_ATTACHCURRENTTHREADASDAEMON_ENTRY(vm, penv, _args);
-  if (vm_created != COMPLETE) {
+  if (vm_created == NOT_CREATED) {
     // Not sure how we could possibly get here.
   HOTSPOT_JNI_ATTACHCURRENTTHREADASDAEMON_RETURN((uint32_t) JNI_ERR);
```
[the remark this answers is not in the document]
Comment (dholmes-ora) [reply src/hotspot/share/prims/jni.cpp]: Fixed

**Comment by dholmes-ora** [comment]
Thanks for the reviews @dcubed-ojdk  and @gerard-ziemski !

I have address the indentation issue and removed the recently added ProblemList entry.

@AlanBateman perhaps we should look further at the "live phase" but I don't think this is inherently tied to JVMTI. In theory the core libs could create a native thread that wants to attach during VM init. Either way I'm reluctant to introduce any change in behaviour other than that which was needed to fix the GetCreatedJavaVMs related crash. But I can file a RFE to look at this.

/integrate

**Pull request closed without merging**